firedrake/adjoint_utils/blocks/solving.py

```diff
@@ -779,7 +786,11 @@ def evaluate_adj_component(self, inputs, adj_inputs, block_variable, idx,
         replace_map[self.func] = self.get_outputs()[0].saved_output
         dFdm = replace(dFdm, replace_map)
 
-        dFdm = dFdm * adj_sol
+        if isinstance(dFdm, firedrake.Argument):
+            #  Corner case. Should be fixed more permanently upstream in UFL.
```

Comment:
Ideally this should have a link to a github issue. Is this the relevant UFL issue: https://github.com/FEniCS/ufl/issues/301? If not then we should open one.

We should also open a Firedrake issue to keep track of the fact that this needs updating once the UFL problem is fixed.